google-cloud-bigquery/src/main/java/com/google/cloud/bigquery/QueryRequest.java

```diff
@@ -102,6 +106,33 @@ public Builder setQuery(String query) {
     }
 
     /**
+     * Adds a QueryParameter to the list of query parameters. See
+     * {@link #setQueryParameters(Iterable)} for more details on the input requirements.
+     */
+    public Builder addQueryParameter(QueryParameter queryParameter) {
+      if (queryParameters == null) {
+        queryParameters = Lists.newArrayList();
+      }
+      queryParameters.add(checkNotNull(queryParameter));
+      return this;
+    }
+
+    /**
+     * Sets the query parameters for the query.
+     *
+     * The set of query parameters must either be all positional or all named parameters.
+     * Positional parameters are denoted in the query with a question mark (?), and named
```

Comment:
Do you think this provides value in this case? An issue is that half of what needs to be done is a textual query, which also has to be consistent with itself and with these parameters, and can't be checked at compile time. Someone could pass the compile check (and even runtime check) for the query parameters, but still have it rejected by the server for inconsistency. 

If we were going to go the route of enforcing consistency here, I might prefer to just hide the concept of QueryParameter and do something here like:

```
public Builder addPositionalParameter(QueryParameterValue value);
public Builder setPositionalParameters(Iterable<QueryParameterValue> values);

public Builder addNamedParameter(String name, QueryParameterValue value);
public Builder setNamedParameters(Map<String, QueryParameterValue> values);
```

How important do you think this is to address?